gamemodes/terrortown/gamemode/shared/sh_shop.lua

```diff
@@ -0,0 +1,320 @@
+		if not RANDOMSHOP[ply] or #RANDOMSHOP[ply] == 0 then
+			return false, shop.statusCode.NORANDOMSHOP
+		end
+
+		local containedInRandomShop = false
+
+		for i = 1, #RANDOMSHOP[ply] do
+			if RANDOMSHOP[ply][i].id ~= equipmentId then continue end
+
+			containedInRandomShop = true
+		end
+
+		if not containedInRandomShop then
+			return false, shop.statusCode.NOTINSHOP
+		end
+	end
+
+	-- Still support old items
+	local oldId = isItem and equipment.oldId or equipmentId
+
+	---
+	-- @note Keep compatibility with old addons
+	-- @realm server
+	if not hook.Run("TTTCanOrderEquipment", ply, oldId, isItem) then
+		return false, shop.statusCode.BLOCKEDBYOLDHOOK
+	end
+
+	---
+	-- @note Add our own hook with more consistent class parameter and some more information
+	-- @realm server
+	local allow, ignoreCost = hook.Run("TTT2CanOrderEquipment", ply, equipmentId, isItem, credits)
+	if not allow then
+		return false, shop.statusCode.BLOCKEDBYTTT2HOOK
+	end
+
+	if ignoreCost then
+		return true, shop.statusCode.SUCCESSIGNORECOST
+	end
+
+	return true, shop.statusCode.SUCCESS
+end
+
+
+---
+-- Buys for player the equipment with the corresponding Id
+-- @param Player ply The player to buy the equipment for
+-- @param string equipmentId The name of the equipment to buy
+-- @return bool True, if equipment can be bought
+-- @return number The shop.statusCode, that lead to the decision
+-- @realm shared
+function shop.BuyEquipment(ply, equipmentId)
+	local isBuyable, statusCode = shop.CanBuyEquipment(ply, equipmentId)
+	if not isBuyable then
+		return false, statusCode
+	end
+
+	if CLIENT then
+		net.Start("TTT2OrderEquipment")
+		net.WriteString(equipmentId)
+		net.SendToServer()
+	else
+		local isItem = items.IsItem(equipmentId)
+
+		-- we use weapons.GetStored to save time making an unnecessary copy, we will not be modifying it
+		local equipment = isItem and items.GetStored(equipmentId) or weapons.GetStored(equipmentId)
+		local credits = equipment.credits or 1
+		local ignoreCost = statusCode == shop.statusCode.SUCCESSIGNORECOST
+
+		if not ignoreCost then
+			ply:SubtractCredits(credits)
+		end
+
+		ply:AddBought(equipmentId)
+
+		-- no longer restricted to only WEAPON_EQUIP weapons, just anything that
+		-- is whitelisted and carryable
+		if isItem then
+			local item = ply:GiveEquipmentItem(equipmentId)
+
+			if isfunction(item.Bought) then
+				item:Bought(ply)
+			end
+		else
+			ply:GiveEquipmentWeapon(equipmentId, function(_ply, _equipmentId, _weapon)
+				if isfunction(_weapon.WasBought) then
+					-- some weapons give extra ammo after being bought, etc
+					_weapon:WasBought(_ply)
+				end
+			end)
+		end
+
+		LANG.Msg(ply, "buy_received", nil, MSG_MSTACK_ROLE)
+
+		timer.Simple(0.5, function()
+			if not IsValid(ply) then return end
+
+			net.Start("TTT_BoughtItem")
+			net.WriteString(equipmentId)
+			net.Send(ply)
+		end)
+
+		if GetGlobalBool("ttt2_random_shop_reroll_per_buy") then
+			shop.RerollShop(ply)
```

Comment:
Made a mistake here, this shouldnt cost any credits and should be forced either way. Will fix it.